_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,133 @@
+        logger.log("Detected fraud transaction {}", t);
+        // Integration with external service
+        alertingService.sendAlert(new FraudTransactionAlert(t));
+      }
+    }
+    ```
+    Another example may be a `BufferedLogger` class which buffers logging statements in memory and manages a separate background thread that takes statements from the buffer and writes them to actual external logger:
+    ```
+    class BufferedLogger implements Logger {
+      private final Logger delegate;
+      private final ExecutorService backgroundWorker;
+      private final BlockingQueue<Statement> buffer;
+
+      BufferedLogger(Logger delegate) {
+        this.delegate = delegate;
+        this.backgroundWorker = newSingleThreadExecutor();
+        this.buffer = new ArrayBlockingQueue<>(100);
+        backgroundWorker.execute(this::writeStatementsInBackground);
+      }
+      
+      @Override public void log(Statement s) {
+        putUninterruptibly(buffer, s);
+      }
+
+      private void writeStatementsInBackground() {
+        while (true) {
+          Statement s = takeUninterruptibly(buffer);
+          delegate.log(s);
+        }
+      }
+    }
+    ```
+   `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.
+ 
+   In a distributed system, examples of services with this type of responsibility could be a proxy, a load balancer, or a service transparently caching responses from or buffering requests to some other service.
+
+#### How small or large a responsibility should be?
+I hope the examples above give some more grounded sense of what a responsibility of a method, class, module, or service could be. However, they still provide no actionable guidance on how finely you should chop responsibilities between the components of your system. For example:
+
+ - Should conversion from XML to JSON be a responsibility of a single method (or a class), or it should be split between two methods: one translates XML into a tree, and another serializes a tree into JSON? Or these should be separate methods belonging to a single class?
+ - Should individual types of interactions with an external service (such as different RPC operations) be responsibilities of separate classes,  or they should all belong to a single class? Or, perhaps, interactions should be grouped, such as read operations going to one class and write operations going to another?
+ - How small parts of control flow should be encapsulated in individual methods?
+ - What kinds of responsibilities should be split between different (micro)services?
+
+Uncle Bob Martin (who first proposed Single Responsibility Principle) suggests that components should be broken down until each one has only one *reason to change*. However,  to me, this criterion still doesn't feel very instructive. Consider the `processTransaction` method above. There may be many reasons to change it:
```

Comment:
When referring to a method, add the parentheses. This makes it clear at a glance that you're talking about a method and not a variable name. I.e. `processTransaction()` instead of `processTransaction`.

Reply:
Fixed